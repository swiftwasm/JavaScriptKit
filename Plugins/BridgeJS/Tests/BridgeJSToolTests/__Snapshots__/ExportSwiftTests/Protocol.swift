// NOTICE: This is auto-generated code by BridgeJS from JavaScriptKit,
// DO NOT EDIT.
//
// To update this file, just rebuild your project or run
// `swift package bridge-js`.

@_spi(BridgeJS) import JavaScriptKit

struct AnyMyViewControllerDelegate: MyViewControllerDelegate, _BridgedSwiftProtocolWrapper {
    let jsObject: JSObject

    func onSomethingHappened() {
    @_extern(wasm, module: "TestModule", name: "bjs_MyViewControllerDelegate_onSomethingHappened")
    func _extern_onSomethingHappened(this: Int32)
    _extern_onSomethingHappened(this: Int32(bitPattern: jsObject.id))
    }

    func onValueChanged(_ value: String) {
        @_extern(wasm, module: "TestModule", name: "bjs_MyViewControllerDelegate_onValueChanged")
        func _extern_onValueChanged(this: Int32, value: Int32)
        _extern_onValueChanged(this: Int32(bitPattern: jsObject.id), value: value.bridgeJSLowerParameter())
    }

    func onCountUpdated(count: Int) -> Bool {
        @_extern(wasm, module: "TestModule", name: "bjs_MyViewControllerDelegate_onCountUpdated")
        func _extern_onCountUpdated(this: Int32, count: Int32) -> Int32
        let ret = _extern_onCountUpdated(this: Int32(bitPattern: jsObject.id), count: count.bridgeJSLowerParameter())
        return Bool.bridgeJSLiftReturn(ret)
    }

    func onLabelUpdated(_ prefix: String, _ suffix: String) {
        @_extern(wasm, module: "TestModule", name: "bjs_MyViewControllerDelegate_onLabelUpdated")
        func _extern_onLabelUpdated(this: Int32, prefix: Int32, suffix: Int32)
        _extern_onLabelUpdated(this: Int32(bitPattern: jsObject.id), prefix: prefix.bridgeJSLowerParameter(), suffix: suffix.bridgeJSLowerParameter())
    }

    func isCountEven() -> Bool {
        @_extern(wasm, module: "TestModule", name: "bjs_MyViewControllerDelegate_isCountEven")
        func _extern_isCountEven(this: Int32) -> Int32
        let ret = _extern_isCountEven(this: Int32(bitPattern: jsObject.id))
        return Bool.bridgeJSLiftReturn(ret)
    }

    var eventCount: Int {
        get {
            @_extern(wasm, module: "TestModule", name: "bjs_MyViewControllerDelegate_eventCount_get")
            func _extern_get(this: Int32) -> Int32
            let ret = _extern_get(this: Int32(bitPattern: jsObject.id))
            return Int.bridgeJSLiftReturn(ret)
        }
        set {
            @_extern(wasm, module: "TestModule", name: "bjs_MyViewControllerDelegate_eventCount_set")
            func _extern_set(this: Int32, value: Int32)
            _extern_set(this: Int32(bitPattern: jsObject.id), value: newValue.bridgeJSLowerParameter())
        }
    }

    var delegateName: String {
        get {
            @_extern(wasm, module: "TestModule", name: "bjs_MyViewControllerDelegate_delegateName_get")
            func _extern_get(this: Int32) -> Int32
            let ret = _extern_get(this: Int32(bitPattern: jsObject.id))
            return String.bridgeJSLiftReturn(ret)
        }
    }

    static func bridgeJSLiftParameter(_ value: Int32) -> Self {
        return AnyMyViewControllerDelegate(jsObject: JSObject(id: UInt32(bitPattern: value)))
    }
}

@_expose(wasm, "bjs_MyViewController_init")
@_cdecl("bjs_MyViewController_init")
public func _bjs_MyViewController_init(delegate: Int32) -> UnsafeMutableRawPointer {
    #if arch(wasm32)
    let ret = MyViewController(delegate: AnyMyViewControllerDelegate.bridgeJSLiftParameter(delegate))
    return ret.bridgeJSLowerReturn()
    #else
    fatalError("Only available on WebAssembly")
    #endif
}

@_expose(wasm, "bjs_MyViewController_triggerEvent")
@_cdecl("bjs_MyViewController_triggerEvent")
public func _bjs_MyViewController_triggerEvent(_self: UnsafeMutableRawPointer) -> Void {
    #if arch(wasm32)
    MyViewController.bridgeJSLiftParameter(_self).triggerEvent()
    #else
    fatalError("Only available on WebAssembly")
    #endif
}

@_expose(wasm, "bjs_MyViewController_updateValue")
@_cdecl("bjs_MyViewController_updateValue")
public func _bjs_MyViewController_updateValue(_self: UnsafeMutableRawPointer, valueBytes: Int32, valueLength: Int32) -> Void {
    #if arch(wasm32)
    MyViewController.bridgeJSLiftParameter(_self).updateValue(_: String.bridgeJSLiftParameter(valueBytes, valueLength))
    #else
    fatalError("Only available on WebAssembly")
    #endif
}

@_expose(wasm, "bjs_MyViewController_updateCount")
@_cdecl("bjs_MyViewController_updateCount")
public func _bjs_MyViewController_updateCount(_self: UnsafeMutableRawPointer, count: Int32) -> Int32 {
    #if arch(wasm32)
    let ret = MyViewController.bridgeJSLiftParameter(_self).updateCount(_: Int.bridgeJSLiftParameter(count))
    return ret.bridgeJSLowerReturn()
    #else
    fatalError("Only available on WebAssembly")
    #endif
}

@_expose(wasm, "bjs_MyViewController_updateLabel")
@_cdecl("bjs_MyViewController_updateLabel")
public func _bjs_MyViewController_updateLabel(_self: UnsafeMutableRawPointer, prefixBytes: Int32, prefixLength: Int32, suffixBytes: Int32, suffixLength: Int32) -> Void {
    #if arch(wasm32)
    MyViewController.bridgeJSLiftParameter(_self).updateLabel(_: String.bridgeJSLiftParameter(prefixBytes, prefixLength), _: String.bridgeJSLiftParameter(suffixBytes, suffixLength))
    #else
    fatalError("Only available on WebAssembly")
    #endif
}

@_expose(wasm, "bjs_MyViewController_checkEvenCount")
@_cdecl("bjs_MyViewController_checkEvenCount")
public func _bjs_MyViewController_checkEvenCount(_self: UnsafeMutableRawPointer) -> Int32 {
    #if arch(wasm32)
    let ret = MyViewController.bridgeJSLiftParameter(_self).checkEvenCount()
    return ret.bridgeJSLowerReturn()
    #else
    fatalError("Only available on WebAssembly")
    #endif
}

@_expose(wasm, "bjs_MyViewController_delegate_get")
@_cdecl("bjs_MyViewController_delegate_get")
public func _bjs_MyViewController_delegate_get(_self: UnsafeMutableRawPointer) -> Int32 {
    #if arch(wasm32)
    let ret = MyViewController.bridgeJSLiftParameter(_self).delegate as! AnyMyViewControllerDelegate
    return ret.bridgeJSLowerReturn()
    #else
    fatalError("Only available on WebAssembly")
    #endif
}

@_expose(wasm, "bjs_MyViewController_delegate_set")
@_cdecl("bjs_MyViewController_delegate_set")
public func _bjs_MyViewController_delegate_set(_self: UnsafeMutableRawPointer, value: Int32) -> Void {
    #if arch(wasm32)
    MyViewController.bridgeJSLiftParameter(_self).delegate = AnyMyViewControllerDelegate.bridgeJSLiftParameter(value)
    #else
    fatalError("Only available on WebAssembly")
    #endif
}

@_expose(wasm, "bjs_MyViewController_secondDelegate_get")
@_cdecl("bjs_MyViewController_secondDelegate_get")
public func _bjs_MyViewController_secondDelegate_get(_self: UnsafeMutableRawPointer) -> Void {
    #if arch(wasm32)
    let ret = MyViewController.bridgeJSLiftParameter(_self).secondDelegate.flatMap {
        $0 as? AnyMyViewControllerDelegate
    }
    return ret.bridgeJSLowerReturn()
    #else
    fatalError("Only available on WebAssembly")
    #endif
}

@_expose(wasm, "bjs_MyViewController_secondDelegate_set")
@_cdecl("bjs_MyViewController_secondDelegate_set")
public func _bjs_MyViewController_secondDelegate_set(_self: UnsafeMutableRawPointer, valueIsSome: Int32, valueValue: Int32) -> Void {
    #if arch(wasm32)
    MyViewController.bridgeJSLiftParameter(_self).secondDelegate = Optional<AnyMyViewControllerDelegate>.bridgeJSLiftParameter(valueIsSome, valueValue)
    #else
    fatalError("Only available on WebAssembly")
    #endif
}

@_expose(wasm, "bjs_MyViewController_deinit")
@_cdecl("bjs_MyViewController_deinit")
public func _bjs_MyViewController_deinit(pointer: UnsafeMutableRawPointer) {
    Unmanaged<MyViewController>.fromOpaque(pointer).release()
}

extension MyViewController: ConvertibleToJSValue, _BridgedSwiftHeapObject {
    var jsValue: JSValue {
        #if arch(wasm32)
        @_extern(wasm, module: "TestModule", name: "bjs_MyViewController_wrap")
        func _bjs_MyViewController_wrap(_: UnsafeMutableRawPointer) -> Int32
        #else
        func _bjs_MyViewController_wrap(_: UnsafeMutableRawPointer) -> Int32 {
            fatalError("Only available on WebAssembly")
        }
        #endif
        return .object(JSObject(id: UInt32(bitPattern: _bjs_MyViewController_wrap(Unmanaged.passRetained(self).toOpaque()))))
    }
}