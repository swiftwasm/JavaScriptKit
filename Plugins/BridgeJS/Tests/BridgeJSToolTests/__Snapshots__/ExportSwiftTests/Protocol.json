{
  "classes" : [
    {
      "constructor" : {
        "abiName" : "bjs_Helper_init",
        "effects" : {
          "isAsync" : false,
          "isStatic" : false,
          "isThrows" : false
        },
        "parameters" : [
          {
            "label" : "value",
            "name" : "value",
            "type" : {
              "int" : {

              }
            }
          }
        ]
      },
      "methods" : [
        {
          "abiName" : "bjs_Helper_increment",
          "effects" : {
            "isAsync" : false,
            "isStatic" : false,
            "isThrows" : false
          },
          "name" : "increment",
          "parameters" : [

          ],
          "returnType" : {
            "void" : {

            }
          }
        }
      ],
      "name" : "Helper",
      "properties" : [
        {
          "isReadonly" : false,
          "isStatic" : false,
          "name" : "value",
          "type" : {
            "int" : {

            }
          }
        }
      ],
      "swiftCallName" : "Helper"
    },
    {
      "constructor" : {
        "abiName" : "bjs_MyViewController_init",
        "effects" : {
          "isAsync" : false,
          "isStatic" : false,
          "isThrows" : false
        },
        "parameters" : [
          {
            "label" : "delegate",
            "name" : "delegate",
            "type" : {
              "swiftProtocol" : {
                "_0" : "MyViewControllerDelegate"
              }
            }
          }
        ]
      },
      "methods" : [
        {
          "abiName" : "bjs_MyViewController_triggerEvent",
          "effects" : {
            "isAsync" : false,
            "isStatic" : false,
            "isThrows" : false
          },
          "name" : "triggerEvent",
          "parameters" : [

          ],
          "returnType" : {
            "void" : {

            }
          }
        },
        {
          "abiName" : "bjs_MyViewController_updateValue",
          "effects" : {
            "isAsync" : false,
            "isStatic" : false,
            "isThrows" : false
          },
          "name" : "updateValue",
          "parameters" : [
            {
              "label" : "_",
              "name" : "value",
              "type" : {
                "string" : {

                }
              }
            }
          ],
          "returnType" : {
            "void" : {

            }
          }
        },
        {
          "abiName" : "bjs_MyViewController_updateCount",
          "effects" : {
            "isAsync" : false,
            "isStatic" : false,
            "isThrows" : false
          },
          "name" : "updateCount",
          "parameters" : [
            {
              "label" : "_",
              "name" : "count",
              "type" : {
                "int" : {

                }
              }
            }
          ],
          "returnType" : {
            "bool" : {

            }
          }
        },
        {
          "abiName" : "bjs_MyViewController_updateLabel",
          "effects" : {
            "isAsync" : false,
            "isStatic" : false,
            "isThrows" : false
          },
          "name" : "updateLabel",
          "parameters" : [
            {
              "label" : "_",
              "name" : "prefix",
              "type" : {
                "string" : {

                }
              }
            },
            {
              "label" : "_",
              "name" : "suffix",
              "type" : {
                "string" : {

                }
              }
            }
          ],
          "returnType" : {
            "void" : {

            }
          }
        },
        {
          "abiName" : "bjs_MyViewController_checkEvenCount",
          "effects" : {
            "isAsync" : false,
            "isStatic" : false,
            "isThrows" : false
          },
          "name" : "checkEvenCount",
          "parameters" : [

          ],
          "returnType" : {
            "bool" : {

            }
          }
        },
        {
          "abiName" : "bjs_MyViewController_sendHelper",
          "effects" : {
            "isAsync" : false,
            "isStatic" : false,
            "isThrows" : false
          },
          "name" : "sendHelper",
          "parameters" : [
            {
              "label" : "_",
              "name" : "helper",
              "type" : {
                "swiftHeapObject" : {
                  "_0" : "Helper"
                }
              }
            }
          ],
          "returnType" : {
            "void" : {

            }
          }
        }
      ],
      "name" : "MyViewController",
      "properties" : [
        {
          "isReadonly" : false,
          "isStatic" : false,
          "name" : "delegate",
          "type" : {
            "swiftProtocol" : {
              "_0" : "MyViewControllerDelegate"
            }
          }
        },
        {
          "isReadonly" : false,
          "isStatic" : false,
          "name" : "secondDelegate",
          "type" : {
            "optional" : {
              "_0" : {
                "swiftProtocol" : {
                  "_0" : "MyViewControllerDelegate"
                }
              }
            }
          }
        }
      ],
      "swiftCallName" : "MyViewController"
    }
  ],
  "enums" : [
    {
      "cases" : [
        {
          "associatedValues" : [

          ],
          "name" : "north"
        },
        {
          "associatedValues" : [

          ],
          "name" : "south"
        },
        {
          "associatedValues" : [

          ],
          "name" : "east"
        },
        {
          "associatedValues" : [

          ],
          "name" : "west"
        }
      ],
      "emitStyle" : "const",
      "name" : "Direction",
      "staticMethods" : [

      ],
      "staticProperties" : [

      ],
      "swiftCallName" : "Direction"
    },
    {
      "cases" : [
        {
          "associatedValues" : [

          ],
          "name" : "test",
          "rawValue" : "test"
        },
        {
          "associatedValues" : [

          ],
          "name" : "test2",
          "rawValue" : "test2"
        }
      ],
      "emitStyle" : "const",
      "name" : "ExampleEnum",
      "rawType" : "String",
      "staticMethods" : [

      ],
      "staticProperties" : [

      ],
      "swiftCallName" : "ExampleEnum"
    },
    {
      "cases" : [
        {
          "associatedValues" : [
            {
              "type" : {
                "string" : {

                }
              }
            }
          ],
          "name" : "success"
        },
        {
          "associatedValues" : [
            {
              "type" : {
                "int" : {

                }
              }
            }
          ],
          "name" : "failure"
        }
      ],
      "emitStyle" : "const",
      "name" : "Result",
      "staticMethods" : [

      ],
      "staticProperties" : [

      ],
      "swiftCallName" : "Result"
    },
    {
      "cases" : [
        {
          "associatedValues" : [

          ],
          "name" : "low",
          "rawValue" : "-1"
        },
        {
          "associatedValues" : [

          ],
          "name" : "medium",
          "rawValue" : "0"
        },
        {
          "associatedValues" : [

          ],
          "name" : "high",
          "rawValue" : "1"
        }
      ],
      "emitStyle" : "const",
      "name" : "Priority",
      "rawType" : "Int",
      "staticMethods" : [

      ],
      "staticProperties" : [

      ],
      "swiftCallName" : "Priority"
    }
  ],
  "functions" : [

  ],
  "moduleName" : "TestModule",
  "protocols" : [
    {
      "methods" : [
        {
          "abiName" : "bjs_MyViewControllerDelegate_onSomethingHappened",
          "effects" : {
            "isAsync" : false,
            "isStatic" : false,
            "isThrows" : false
          },
          "name" : "onSomethingHappened",
          "parameters" : [

          ],
          "returnType" : {
            "void" : {

            }
          }
        },
        {
          "abiName" : "bjs_MyViewControllerDelegate_onValueChanged",
          "effects" : {
            "isAsync" : false,
            "isStatic" : false,
            "isThrows" : false
          },
          "name" : "onValueChanged",
          "parameters" : [
            {
              "label" : "_",
              "name" : "value",
              "type" : {
                "string" : {

                }
              }
            }
          ],
          "returnType" : {
            "void" : {

            }
          }
        },
        {
          "abiName" : "bjs_MyViewControllerDelegate_onCountUpdated",
          "effects" : {
            "isAsync" : false,
            "isStatic" : false,
            "isThrows" : false
          },
          "name" : "onCountUpdated",
          "parameters" : [
            {
              "label" : "count",
              "name" : "count",
              "type" : {
                "int" : {

                }
              }
            }
          ],
          "returnType" : {
            "bool" : {

            }
          }
        },
        {
          "abiName" : "bjs_MyViewControllerDelegate_onLabelUpdated",
          "effects" : {
            "isAsync" : false,
            "isStatic" : false,
            "isThrows" : false
          },
          "name" : "onLabelUpdated",
          "parameters" : [
            {
              "label" : "_",
              "name" : "prefix",
              "type" : {
                "string" : {

                }
              }
            },
            {
              "label" : "_",
              "name" : "suffix",
              "type" : {
                "string" : {

                }
              }
            }
          ],
          "returnType" : {
            "void" : {

            }
          }
        },
        {
          "abiName" : "bjs_MyViewControllerDelegate_isCountEven",
          "effects" : {
            "isAsync" : false,
            "isStatic" : false,
            "isThrows" : false
          },
          "name" : "isCountEven",
          "parameters" : [

          ],
          "returnType" : {
            "bool" : {

            }
          }
        },
        {
          "abiName" : "bjs_MyViewControllerDelegate_onHelperUpdated",
          "effects" : {
            "isAsync" : false,
            "isStatic" : false,
            "isThrows" : false
          },
          "name" : "onHelperUpdated",
          "parameters" : [
            {
              "label" : "_",
              "name" : "helper",
              "type" : {
                "swiftHeapObject" : {
                  "_0" : "Helper"
                }
              }
            }
          ],
          "returnType" : {
            "void" : {

            }
          }
        },
        {
          "abiName" : "bjs_MyViewControllerDelegate_createHelper",
          "effects" : {
            "isAsync" : false,
            "isStatic" : false,
            "isThrows" : false
          },
          "name" : "createHelper",
          "parameters" : [

          ],
          "returnType" : {
            "swiftHeapObject" : {
              "_0" : "Helper"
            }
          }
        },
        {
          "abiName" : "bjs_MyViewControllerDelegate_onOptionalHelperUpdated",
          "effects" : {
            "isAsync" : false,
            "isStatic" : false,
            "isThrows" : false
          },
          "name" : "onOptionalHelperUpdated",
          "parameters" : [
            {
              "label" : "_",
              "name" : "helper",
              "type" : {
                "optional" : {
                  "_0" : {
                    "swiftHeapObject" : {
                      "_0" : "Helper"
                    }
                  }
                }
              }
            }
          ],
          "returnType" : {
            "void" : {

            }
          }
        },
        {
          "abiName" : "bjs_MyViewControllerDelegate_createOptionalHelper",
          "effects" : {
            "isAsync" : false,
            "isStatic" : false,
            "isThrows" : false
          },
          "name" : "createOptionalHelper",
          "parameters" : [

          ],
          "returnType" : {
            "optional" : {
              "_0" : {
                "swiftHeapObject" : {
                  "_0" : "Helper"
                }
              }
            }
          }
        },
        {
          "abiName" : "bjs_MyViewControllerDelegate_createEnum",
          "effects" : {
            "isAsync" : false,
            "isStatic" : false,
            "isThrows" : false
          },
          "name" : "createEnum",
          "parameters" : [

          ],
          "returnType" : {
            "rawValueEnum" : {
              "_0" : "ExampleEnum",
              "_1" : "String"
            }
          }
        },
        {
          "abiName" : "bjs_MyViewControllerDelegate_handleResult",
          "effects" : {
            "isAsync" : false,
            "isStatic" : false,
            "isThrows" : false
          },
          "name" : "handleResult",
          "parameters" : [
            {
              "label" : "_",
              "name" : "result",
              "type" : {
                "associatedValueEnum" : {
                  "_0" : "Result"
                }
              }
            }
          ],
          "returnType" : {
            "void" : {

            }
          }
        },
        {
          "abiName" : "bjs_MyViewControllerDelegate_getResult",
          "effects" : {
            "isAsync" : false,
            "isStatic" : false,
            "isThrows" : false
          },
          "name" : "getResult",
          "parameters" : [

          ],
          "returnType" : {
            "associatedValueEnum" : {
              "_0" : "Result"
            }
          }
        }
      ],
      "name" : "MyViewControllerDelegate",
      "properties" : [
        {
          "isReadonly" : false,
          "name" : "eventCount",
          "type" : {
            "int" : {

            }
          }
        },
        {
          "isReadonly" : true,
          "name" : "delegateName",
          "type" : {
            "string" : {

            }
          }
        },
        {
          "isReadonly" : false,
          "name" : "optionalName",
          "type" : {
            "optional" : {
              "_0" : {
                "string" : {

                }
              }
            }
          }
        },
        {
          "isReadonly" : false,
          "name" : "optionalRawEnum",
          "type" : {
            "optional" : {
              "_0" : {
                "rawValueEnum" : {
                  "_0" : "ExampleEnum",
                  "_1" : "String"
                }
              }
            }
          }
        },
        {
          "isReadonly" : false,
          "name" : "rawStringEnum",
          "type" : {
            "rawValueEnum" : {
              "_0" : "ExampleEnum",
              "_1" : "String"
            }
          }
        },
        {
          "isReadonly" : false,
          "name" : "result",
          "type" : {
            "associatedValueEnum" : {
              "_0" : "Result"
            }
          }
        },
        {
          "isReadonly" : false,
          "name" : "optionalResult",
          "type" : {
            "optional" : {
              "_0" : {
                "associatedValueEnum" : {
                  "_0" : "Result"
                }
              }
            }
          }
        },
        {
          "isReadonly" : false,
          "name" : "direction",
          "type" : {
            "caseEnum" : {
              "_0" : "Direction"
            }
          }
        },
        {
          "isReadonly" : false,
          "name" : "directionOptional",
          "type" : {
            "optional" : {
              "_0" : {
                "caseEnum" : {
                  "_0" : "Direction"
                }
              }
            }
          }
        },
        {
          "isReadonly" : false,
          "name" : "priority",
          "type" : {
            "rawValueEnum" : {
              "_0" : "Priority",
              "_1" : "Int"
            }
          }
        },
        {
          "isReadonly" : false,
          "name" : "priorityOptional",
          "type" : {
            "optional" : {
              "_0" : {
                "rawValueEnum" : {
                  "_0" : "Priority",
                  "_1" : "Int"
                }
              }
            }
          }
        }
      ]
    }
  ]
}